camviewer_ui_impl.py

```diff
@@ -2789,3 +2713,100 @@ def getConfig(self):
+        roi = gui.ui.display_image.rectRoi.abs()
+        fd.write(
+            "ROI         %d %d %d %d\n" % (roi.x(), roi.y(), roi.width(), roi.height())
+        )
+        fd.write("globmarks   " + str(int(gui.useglobmarks)) + "\n")
+        fd.write("globmarks2  " + str(int(gui.useglobmarks2)) + "\n")
+        lMarker = gui.ui.display_image.lMarker
+        for i in range(4):
+            fd.write(
+                "m%d          %d %d\n"
+                % (i + 1, lMarker[i].abs().x(), lMarker[i].abs().y())
+            )
+        fd.write("projroi     " + str(int(gui.ui.checkBoxProjRoi.isChecked())) + "\n")
+        fd.write(
+            "projlineout "
+            + str(int(gui.ui.checkBoxM1Lineout.isChecked()))
+            + " "
+            + str(int(gui.ui.checkBoxM2Lineout.isChecked()))
+            + " "
+            + str(int(gui.ui.checkBoxM3Lineout.isChecked()))
+            + " "
+            + str(int(gui.ui.checkBoxM4Lineout.isChecked()))
+            + "\n"
+        )
+        fd.write("projfit     " + str(int(gui.ui.checkBoxFits.isChecked())) + "\n")
+        fd.write(
+            "projfittype "
+            + str(int(gui.ui.radioGaussian.isChecked()))
+            + " "
+            + str(int(gui.ui.radioSG4.isChecked()))
+            + " "
+            + str(int(gui.ui.radioSG6.isChecked()))
+            + "\n"
+        )
+        fd.write("projconstant " + str(int(gui.ui.checkBoxConstant.isChecked())) + "\n")
+        fd.write("projcalib   %g\n" % gui.calib)
+        fd.write('projcalibPV "%s"\n' % gui.calibPVName)
+        fd.write('projdisplayFormat "%s"\n' % gui.displayFormat)
+
+
+def write_global_config(gui: GraphicUserInterface) -> None:
+    with atomic_writer(gui.cfgdir + "GLOBAL") as fd:
+        fd.write("config      " + str(int(gui.ui.showconf.isChecked())) + "\n")
+        fd.write("projection  " + str(int(gui.ui.showproj.isChecked())) + "\n")
+        fd.write("markers     " + str(int(gui.ui.showmarker.isChecked())) + "\n")
+        fd.write("dispspec    " + str(gui.dispspec) + "\n")
+
+
+@contextlib.contextmanager
+def atomic_writer(path: str) -> typing.Iterator[typing.TextIO]:
+    with tempfile.NamedTemporaryFile("w", delete=False) as fd:
+        try:
+            yield fd
+        except Exception as exc:
+            # There is some issue and the temp file is not complete.
+            # Avoid the else block, we don't want to keep the corrupt file.
+            # Show some error instead of bricking the gui
+            print(f"Error writing {path}: {exc}")
+        else:
+            # File must be closed before we can chmod and move it
+            fd.close()
+            # Set -rw-r--r-- instead of temp file default -rw-------
+            os.chmod(fd.name, 0o644)
+            shutil.move(fd.name, path)
+    # If the tempfile still exists, we should clean it up.
+    if os.path.exists(fd.name):
+        os.remove(fd.name)
```

Comment:
Does python's `tempfile` not take care of its own cleanup?  Could we set `delete=True, delete_on_close=False` here so that we can close/move the file, but let the context manager handle deleting it?

Reply:
`delete_on_close` was added in python 3.12 :fire: